object_intern.go

```diff
@@ -97,74 +93,45 @@ func (oi *ObjectIntern) DecompressString(in string) (string, error) {
 	return string(b), nil
 }
 
-// AddOrGet finds or adds and then returns a uintptr to an object and nil.
-// This method assumes that compression is turned off.
+// AddOrGet finds or adds an object and returns its uintptr and nil upon success.
+// This method takes a []byte of the object, and two bools. If compressed
+// is set to true then we assume compression is being used. If safe is set to true
+// then this method will create a copy of the []byte before performing any operations
+// that might modify the backing array.
 // On failure it returns 0 and an error
 //
 // If the object is found in the store its reference count is increased by 1.
 // If the object is added to the store its reference count is set to 1.
-func (oi *ObjectIntern) AddOrGet(obj []byte) (uintptr, error) {
+func (oi *ObjectIntern) AddOrGet(obj []byte, compressed, safe bool) (uintptr, error) {
 	var addr gos.ObjAddr
 	var ok bool
 	var err error
 	var objString string
-
-	// create a copy so we don't modify the original, especially because compression is turned off
-	// when compression is turned on using SHOCO it will return a new slice, which is not the case here
-	objComp := make([]byte, len(obj))
-	copy(objComp, obj)
-	objString = string(obj)
-
-	// acquire lock
-	oi.Lock()
-
-	// try to find the object in the index
-	addr, ok = oi.ObjIndex[objString]
-	if ok {
-		// increment reference count by 1
-		(*(*uint32)(unsafe.Pointer(addr + uintptr(len(objComp)))))++
-		oi.Unlock()
-		return addr, nil
-	}
-
-	// The object is not in the index therefore it is not in the store.
-	// We need to set its initial reference count to 1 before adding it
-	objComp = append(objComp, []byte{0x1, 0x0, 0x0, 0x0}...)
-	addr, err = oi.Store.Add(objComp)
-	if err != nil {
-		oi.Unlock()
-		return 0, err
+	var objComp []byte
+
+	// is the data supposed to be compressed?
+	if compressed {
+		// even so, is compression actually turned on, and do we need to be safe?
+		if oi.conf.Compression == None && safe {
```

Comment:
that's a bit confusing, but i think i understand:
the ObjectIntern - config allows the user to specify a compression method. then on AddOrGet the user can decide whether to compress the value or not, but if the user chooses to compress and no method has been chosen then it will just end up uncompressed anyway. wouldn't it be better to return an error in such a case, otherwise a user might think all values are compressed when they are actually not

Reply:
Yeah, I am removing all of the compressed bools. It is set in the config and can't be modified without creating a new instance of the interning library.